### DISCUSSION TITLE
blog/2023-10-12-memory-cache/

### DISCUSSION THREAD

**Discussion by giscus** [Tom910/personal-blog#16] [General]
# blog/2023-10-12-memory-cache/

Cache is a crucial component of many systems. It helps reduce resource consumption and improves the timing of our systems. Often, I've observed the Node.js community focusing on the speed of cache algorithms. However, a cache algorithm primarily has two parameters - hit rate and speed.

https://amarchenko.dev/blog/2023-10-12-memory-cache/

**Comment by 1a1a11a** [comment]
disclaimer: this is the author of S3-FIFO algorithm

Nice writeup! I really like your blog. 

One suggestion on the result, I believe the traces used in the evaluation is too small and some of them are outdated. When the cache size (< 1 million entries) is small, the small queue is too small and can hurt the miss ratio. I believe most production systems should be able to store millions to billions of objects. 


Second, I believe the results can be interpreted differently. Miss ratio is a more helpful metric because it correlates with the load on the backend and the mean data access latency. 
Therefore, when the hit ratio increase from 98% to 99%, it is actually very large improvement, because half of the misses now become cache hits. 

Happy to chat more. :)

**Reply by Tom910** [reply]
Hi @1a1a11a , thanks for this comment and your original work. 

>One suggestion on the result, I believe the traces used in the evaluation is too small and some of them are outdated. When the cache size (< 1 million entries) is small, the small queue is too small and can hurt the miss ratio. I believe most production systems should be able to store millions to billions of objects.

I agree with you. Maybe do you know any examples of open data sets? I took some of data sets from  https://github.com/1a1a11a/libCacheSim/tree/develop/data but they only have limitted size. Also I found one of solution which can help to generate traces from https://github.com/scalalang2/golang-fifo but I don't sure how good this traces will be

>Second, I believe the results can be interpreted differently. Miss ratio is a more helpful metric because it correlates with the load on the backend and the mean data access latency.
Therefore, when the hit ratio increase from 98% to 99%, it is actually very large improvement, because half of the misses now become cache hits.

I didn't think from that perspective. Yeah, this is valuable comment. Maybe I will change this result after I will try a new approach from your team https://cachemon.github.io/SIEVE-website/ and https://jasony.me/publication/nsdi24-SIEVE.pdf

**Reply by 1a1a11a** [reply]
Hi @Tom910, Greetings! 
Thank you for your reply! 

Zipf (power law) is a common workload model for web workloads, so many existing benchmarks use it, e.g., YCSB. 
Here are some open-source datasets 
* Twitter (KV) https://github.com/twitter/cache-trace, I am the author who released this dataset, and cluster52 is the most important trace in this dataset
* Meta (KV and CDN) https://cachelib.org/docs/Cache_Library_User_Guides/Cachebench_FB_HW_eval#list-of-traces
* Wiki (CDN) https://wikitech.wikimedia.org/wiki/Analytics/Data_Lake/Traffic/Caching 
* More traces (but in a binary format) https://github.com/Thesys-lab/sosp23-s3fifo?tab=readme-ov-file#open-source-cache-traces-and-the-people-behind-them 

Thank you for your interest in SIEVE! There are a few ongoing discussions here and there, and I will share them in a bit.
